### PR TITLE
Avoid completion handler allocation

### PR DESCRIPTION
**Pull request by sithhell**
## Proposed Changes

This patch removes `compose_cb` in favor of `boost::container::small_vector<callback_type, 3>`. The main motivation is to avoid memory allocations for cases where two or more completion handlers are attached to a future.

Flyby: avoid reference counting for pack traversals.

### CODE REVIEW

File: hpx/lcos/detail/future_data.hpp [hunk 1]
```diff
@@ -194,75 +196,6 @@ namespace detail
         std::size_t& count_;
```
Comment (hkaiser) [inline hpx/lcos/detail/future_data.hpp]: I think the `handle_continuation_recursion_count` is not needed anymore.

Comment (hkaiser) [reply hpx/lcos/detail/future_data.hpp]: I take that back, sorry.

Comment (hkaiser) [reply hpx/lcos/detail/future_data.hpp]: But it can be moved to the source file now (out of the header) ...

Comment (sithhell) [reply hpx/lcos/detail/future_data.hpp]: good idea.

**Pull request merged**